### ISSUE TITLE
[runtimes] Root bone scaling affects animation timing

### ISSUE THREAD

**Issue by badlogic** [EsotericSoftware/spine-runtimes#664]
See http://esotericsoftware.com/forum/Scaling-Bug-6820?p=31500#p31500

**Comment by badlogic** [comment]
Can kind of reproduce it with the vine sample in sfml. If i set `root->scaleX/scaleY` != 1, then the animation appears to start later.

**Comment by badlogic** [comment]
Can reproduce with reference implementation. Take `vine` sample from runtimes repo, plug it into this (adapted from `SimpleTest3.java`).

```
/******************************************************************************
 * Spine Runtimes Software License
 * Version 2.3
 * 
 * Copyright (c) 2013-2015, Esoteric Software
 * All rights reserved.
 * 
 * You are granted a perpetual, non-exclusive, non-sublicensable and
 * non-transferable license to use, install, execute and perform the Spine
 * Runtimes Software (the "Software") and derivative works solely for personal
 * or internal use. Without the written permission of Esoteric Software (see
 * Section 2 of the Spine Software License Agreement), you may not (a) modify,
 * translate, adapt or otherwise create derivative works, improvements of the
 * Software or develop new applications using the Software or (b) remove,
 * delete, alter or obscure any trademarks or any copyright, trademark, patent
 * or other intellectual property or proprietary rights notices on or in the
 * Software, including any copy thereof. Redistributions in binary or source
 * form must include this license and terms.
 * 
 * THIS SOFTWARE IS PROVIDED BY ESOTERIC SOFTWARE "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL ESOTERIC SOFTWARE BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

package com.esotericsoftware.spine;

import com.badlogic.gdx.ApplicationAdapter;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.backends.lwjgl.LwjglApplication;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.PolygonSpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;

public class SimpleTest3 extends ApplicationAdapter {
	OrthographicCamera camera;
	PolygonSpriteBatch batch;
	SkeletonMeshRenderer renderer;
	SkeletonRendererDebug debugRenderer;

	TextureAtlas atlas;
	Skeleton skeleton;
	AnimationState state;

	public void create () {
		camera = new OrthographicCamera();
		batch = new PolygonSpriteBatch(); // Required to render meshes. SpriteBatch can't render meshes.
		renderer = new SkeletonMeshRenderer();
		renderer.setPremultipliedAlpha(false);
		debugRenderer = new SkeletonRendererDebug();
		debugRenderer.setMeshTriangles(false);
		debugRenderer.setRegionAttachments(false);
		debugRenderer.setMeshHull(false);

		atlas = new TextureAtlas(Gdx.files.internal("vine/vine.atlas"));
		SkeletonJson json = new SkeletonJson(atlas); // This loads skeleton JSON data, which is stateless.
		json.setScale(0.5f); // Load the skeleton at 50% the size it was in Spine.
		SkeletonData skeletonData = json.readSkeletonData(Gdx.files.internal("vine/vine.json"));

		skeleton = new Skeleton(skeletonData); // Skeleton holds skeleton state (bone positions, slot attachments, etc).
		skeleton.setPosition(320, 20);
		skeleton.getRootBone().scaleY = 0.5f;
		
		AnimationStateData stateData = new AnimationStateData(skeletonData); // Defines mixing (crossfading) between animations.

		state = new AnimationState(stateData); // Holds the animation state for a skeleton (current animation, time, etc).
//		state.setTimeScale(0.6f); // Slow all animations down to 60% speed.

		// Queue animations on tracks 0 and 1.
		state.setAnimation(0, "animation", true);
//		state.setAnimation(1, "empty", false);
//		state.addAnimation(1, "gungrab", false, 2); // Keys in higher tracks override the pose from lower tracks.
		
		state.update(1.0f);
	}

	public void render () {
		 // Update the animation time.

		Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

		state.apply(skeleton); // Poses skeleton using current animations. This sets the bones' local SRT.
		skeleton.updateWorldTransform(); // Uses the bones' local SRT to compute their world SRT.

		// Configure the camera, SpriteBatch, and SkeletonRendererDebug.
		camera.update();
		batch.getProjectionMatrix().set(camera.combined);
		debugRenderer.getShapeRenderer().setProjectionMatrix(camera.combined);

		batch.begin();
		renderer.draw(batch, skeleton); // Draw the skeleton images.
		batch.end();

		// debugRenderer.draw(skeleton); // Draw debug lines.
	}

	public void resize (int width, int height) {
		camera.setToOrtho(false); // Update camera with new size.
	}

	public void dispose () {
		atlas.dispose();
	}

	public static void main (String[] args) throws Exception {
		new LwjglApplication(new SimpleTest3());
	}
}
```

The smaller `root.scaleY` the longer it takes for the animation to start.

**Comment by pharan** [comment]
Tested in Unity and it happened, but I had a better hunch about what was going on. So I opened vine.spine.

So, more info:
In the vine skeleton, the vine-path PathConstraint [spacing is set to Length](http://esotericsoftware.com/spine-path-constraints#Spacing-mode). The resolved position values depend on the length of the bones. And the way @Shiu animated it, the PathConstraint spacing amount is animated from -100 to 20.

When the root bone is scaled, [those bone lengths are affected](https://github.com/EsotericSoftware/spine-runtimes/blob/master/spine-csharp/src/PathConstraint.cs#L93-L97). But the spacing between them is [clamped at 0](https://github.com/EsotericSoftware/spine-runtimes/blob/master/spine-csharp/src/PathConstraint.cs#L97).

I think that means shorter bones (caused by rootBone.scaleY) causes the resolved spacing to take longer to reach a positive value and become visible.

So this is just a sideffect of using negative values in Length spacing mode. (@NathanSweet ?)

PS
btw, the vine atlas exports in the repo are incorrect.
It failed to load because the path in the skeleton json didn't match the path of the atlas. (skeleton's attachment path is "images/vine", atlas's region path just says "vine"). You're exporting this with a batch file, right?

**Comment by NathanSweet** [comment]
Good stuff @pharan, thanks. Your assessment looks right. I've committed a [new vine project](https://github.com/EsotericSoftware/spine-runtimes/raw/master/examples/vine/vine.spine) that doesn't use path timelines, instead it uses scale on the `vine1` bone. I also fixed its images path to use the `images` folder. Nice catch!

**Comment by NathanSweet** [comment]
BTW, I confused myself briefly while messing with the vine project (before my updated version). If I change the spacing mode to `fixed` then the spacing keys are still negative but when applied get clamped to zero, as fixed doesn't allow negative spacing.

**Comment by chadlayton** [comment]
Hey, following up from the forum thread. Sorry, I am failing to be clear. The problem for me wasn't specific to the vine example - I just modified the vine example to sanity check my animation. I also misinterpreted the problem @badlogic described above as being the same as mine. It is not. Let me try again: 

I attached my project to my latest post to the forum thread. I'd attach it here too but GitHub isn't allow zip files for some reason. Maybe this link will work: http://esotericsoftware.com/forum/download/file.php?id=2522

Here is what the animation looks like, correctly, in the editor: https://gfycat.com/SeriousPeskyHedgehog

And here is a test function:

```cpp
void spine_test(const char* atlas_file, const char* skeleton_file, const char* animation_name, float rotation)
{
	// Load atlas, skeleton, and animations.
	spAtlas* atlas = spAtlas_createFromFile(atlas_file, 0);
	spSkeletonJson* json = spSkeletonJson_create(atlas);
	json->scale = 0.5f;
	spSkeletonData *skeletonData = spSkeletonJson_readSkeletonDataFile(json, skeleton_file);
	if (!skeletonData) {
		printf("Error: %s\n", json->error);
		exit(0);
	}
	spSkeletonJson_dispose(json);

	SkeletonDrawable* drawable = new SkeletonDrawable(skeletonData);
	drawable->timeScale = 1;

	spSkeleton* skeleton = drawable->skeleton;
	skeleton->x = 320;
	skeleton->y = 590;
	skeleton->root->rotation = rotation;
	spSkeleton_updateWorldTransform(skeleton);

	spAnimationState_setAnimationByName(drawable->state, 0, animation_name, true);

	sf::RenderWindow window(sf::VideoMode(640, 640), "Spine SFML - test");
	window.setFramerateLimit(60);
	sf::Event event;
	sf::Clock deltaClock;
	while (window.isOpen()) {
		while (window.pollEvent(event))
			if (event.type == sf::Event::Closed) window.close();

		float dt = deltaClock.getElapsedTime().asSeconds();
		deltaClock.restart();

		drawable->OnUpdate(dt);

		window.clear();
		window.draw(*drawable);
		window.display();
	}

	spSkeletonData_dispose(skeletonData);
	spAtlas_dispose(atlas);
}
```

If I run the test with no rotation like this:

    spine_test("Hero/Hero.atlas", "Hero/Hero.json", "Idle", 0);

This is the result (note the apparent chop in the playback - not an artifact of the gif): https://gfycat.com/SpitefulEagerIrukandjijellyfish

If instead I run the test with a rotation:

    spine_test("Hero/Hero.atlas", "Hero/Hero.json", "Idle", 10);

The playback appears smoother: https://gfycat.com/PessimisticValuableKoalabear. Though perhaps the problem is still there and simply less noticeable when the sprite is rotated.

I tried animating the rotation and translation in a similar manner and there did not appear to be any problems. So it appears specific to animating the scale.

**Comment by badlogic** [comment]
Thanks for providing a repo case. I've opened a new issue here https://github.com/EsotericSoftware/spine-runtimes/issues/667

**Comment by badlogic** [comment]
I've reproduced this in both spine-sfml as well as the editor. I've traced all calculations down to the vertex level and all of them are correct. I'm pretty certain this is an optical artifact. As soon as you rotate, you get aliasing at the edges, which makes the animation appear more lively.

Here's the sample of yours at the same scale in spine-sfml and the editor:

![http://www.giphy.com/gifs/26BRR32wTpt7uhzC8](http://www.giphy.com/gifs/26BRR32wTpt7uhzC8)

I've put in a different region attachment at the same scale, and that looks a lot smoother, giving some support to my optical illusion theory:

![http://www.giphy.com/gifs/26hitVUCWB9meZAHe](http://www.giphy.com/gifs/26hitVUCWB9meZAHe)

So, since the calculations are correct, and since using a different, less uniform image gets rid of the effect (as well as adding aliasing via rotation) I'm afraid there's not much we can do. Other than fixing everyone's eyes and brains!
